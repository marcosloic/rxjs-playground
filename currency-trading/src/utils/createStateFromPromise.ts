import {
    combineLatest,
    ConnectableObservable,
    from,
    Observable,
    of
} from 'rxjs';
import {
    catchError,
    map,
    publishReplay,
    startWith,
    switchMap,
    switchMapTo,
    take
} from 'rxjs/operators';

export type AsyncState<T> = {
    data: T | null;
    isLoading: boolean;
    error: string | null;
};

export const createStateFromPromise = <T>(
    request: (base: string, quote: string) => Promise<T>,
    firstArg$: Observable<string>,
    secondArg$: Observable<string>,
    loadIntent$: Observable<any>
): Observable<AsyncState<T>> => {
    const argumentStream$ = combineLatest([firstArg$, secondArg$]).pipe(
        take(1)
    );
    const sink$ = loadIntent$.pipe(
        switchMapTo(argumentStream$),
        switchMap(data => {
            const [a, b] = data;
            return from(request(a, b)).pipe(
                map((value: any) => {
                    return {
                        data: value,
                        isLoading: false,
                        error: null
                    };
                }),
                startWith({
                    data: null,
                    isLoading: true,
                    error: null
                }),
                catchError((error: string) => {
                    return of({
                        data: null,
                        isLoading: false,
                        error
                    });
                })
            );
        }),
        publishReplay(1)
    );
    (sink$ as ConnectableObservable<AsyncState<T>>).connect();
    return sink$;
};
