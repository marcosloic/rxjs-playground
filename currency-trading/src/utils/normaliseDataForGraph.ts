import { DailyTimeSeries } from '../types/DailyTimeSeries';

export type GraphData = {
    x: number;
    y: string;
};
export const normaliseDataForGraph = (data: DailyTimeSeries): GraphData[] => {
    const target = Object.entries(data.series);
    const lastMonth = target.splice(0, 29);
    return lastMonth.reduce(
        (acc, val, idx) => {
            const [, value] = val;
            const res = { x: idx + 1, y: value.close };
            acc.push(res);
            return acc;
        },
        [] as GraphData[]
    );
};
