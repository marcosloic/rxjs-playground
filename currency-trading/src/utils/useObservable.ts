import { useState, useLayoutEffect } from 'react';
import { Observable } from 'rxjs';

export const useObservable = <T>(
    observable$: Observable<T>,
    initialValue?: any
): T => {
    const [value, update] = useState(initialValue);

    useLayoutEffect(() => {
        const s = observable$.subscribe(update);
        return () => s.unsubscribe();
    }, [observable$]);

    return value;
};
