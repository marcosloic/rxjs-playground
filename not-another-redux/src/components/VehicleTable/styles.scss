.vehicleTable {
    display: flex;
    flex-flow: row wrap;
    width: 80%;
    margin: 10px auto;
    align-content: center;
    @media screen and (max-width: 992px) {
        width: 95%;
    }
}
