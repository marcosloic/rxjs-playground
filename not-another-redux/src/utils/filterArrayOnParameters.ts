import { ApiData } from '../types/ApiData.interface';

export const filterArrayOnParameters = (
    dataArray: ApiData[],
    parameters: { [k: string]: string }
): ApiData[] => {
    const parametersArray = Object.entries(parameters);

    return dataArray.filter(entry => {
        const isValid = parametersArray.reduce((acc, val) => {
            const [key, filterVal] = val;
            const target = entry[key];
            if (Array.isArray(target)) {
                return acc && target.indexOf(filterVal) !== -1;
            }
            return acc && entry[key] === filterVal;
        }, true);
        return isValid;
    });
};
