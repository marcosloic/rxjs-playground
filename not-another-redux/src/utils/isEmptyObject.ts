// Can be tripped up by Date objects without the constructor check
// https://stackoverflow.com/questions/679915/how-do-i-test-for-an-empty-javascript-object

export const isEmptyObject = obj =>
    Object.entries(obj).length === 0 && obj.constructor === Object;
