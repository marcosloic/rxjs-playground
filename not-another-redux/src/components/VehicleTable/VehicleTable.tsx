import React from 'react';
import { Vehicle } from '../Vehicle/Vehicle';
import './styles.scss';
import { useObservable } from '../../utils/useObservable';
import { Observable } from 'rxjs';
import { ApiData } from '../../types/ApiData.interface';

type VehicleTablePropTypes = {
    tableData$: Observable<ApiData[]>;
};

export const VehicleTable: React.FunctionComponent<
    VehicleTablePropTypes
> = React.memo(props => {
    const { tableData$ } = props;
    const tableData = useObservable(tableData$);
    if (!tableData || tableData.length === 0) {
        return null;
    }
    return (
        <div className={'vehicleTable'}>
            {tableData.map(data => (
                <Vehicle key={data.id} details={data} />
            ))}
        </div>
    );
});

VehicleTable.displayName = 'VehicleTable';
