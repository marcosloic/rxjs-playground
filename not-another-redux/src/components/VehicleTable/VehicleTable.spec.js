import React from 'react';
import TestRenderer from 'react-test-renderer';
import { VehicleTable } from './VehicleTable';
import { useObservable } from '../../utils/useObservable';

jest.mock('../../utils/useObservable');

const tableData = [{ id: 1 }, { id: 2 }];
// Not the most exciting tests but these have no logic at all
describe('The Vehicle component', () => {
    beforeEach(() => {
        useObservable.mockReturnValue(tableData);
    });
    it('Matches the snapshot', () => {
        const vehicleTable = TestRenderer.create(<VehicleTable />);
        expect(vehicleTable.toJSON()).toMatchSnapshot();
    });

    it('Renders vehicles data', () => {
        const vehicleTable = TestRenderer.create(<VehicleTable />).toJSON();

        expect(vehicleTable.children.length).toBe(tableData.length);
    });
});
