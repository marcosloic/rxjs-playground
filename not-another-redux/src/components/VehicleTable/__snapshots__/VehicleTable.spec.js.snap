// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Vehicle component Matches the snapshot 1`] = `
<div
  className="vehicleTable"
>
  <article
    className="vehicle"
  >
    <img />
    <div
      className="content"
    >
      <p>
        <b>
          Brand:
        </b>
         
      </p>
      <p>
        <b>
          Type:
        </b>
         
      </p>
      <p>
        <b>
          Available in:
        </b>
         
      </p>
    </div>
  </article>
  <article
    className="vehicle"
  >
    <img />
    <div
      className="content"
    >
      <p>
        <b>
          Brand:
        </b>
         
      </p>
      <p>
        <b>
          Type:
        </b>
         
      </p>
      <p>
        <b>
          Available in:
        </b>
         
      </p>
    </div>
  </article>
</div>
`;
